Add scoring logic

PeterJCLaw opened the pull request:
Fixes https://github.com/srobo/tasks/issues/471.
Fixes https://github.com/srobo/tasks/issues/470.

PeterJCLaw committed:
Add initial plan for scoring data format

We might switch to having the scorers tally the tokens by type
rather than writing letters for each one, however that should
be fairly simple to switch over to.

PeterJCLaw committed:
Implement scoring for the SR2020 game

RealOrangeOne commented on scoring/score.py:
```diff
@@ -1,14 +1,54 @@
+        def zone_score(zone_data):
+            if not zone_data:
+                return 0
+            return  3 * sum(zone_data.values()) / len(zone_data)
```
I don't think this is quite correct.

> If tokens of both colours are present in a scoring zone,  the value of each token in that scoring zone is reduced from 3 to 1.

This looks to half the points at the end, making the tokens worth 1.5 points rather than 1 point.

eg https://github.com/srobo/sr2020-comp/pull/1/files#diff-e58af9df2705fcda0714a31abe6a7985R141 should be 2 points rather than 3.

PeterJCLaw replied:
Ah, good spot. This is a failing in my recollection of the rules. I even had the rulebook open when I was doing this! :man_facepalming:

PeterJCLaw replied:
c44f3dc

PeterJCLaw committed:
Correct the per-token point value when there's more than one type

An early version of the rules halved the number of points, but that
isn't actually what we went with in the end.

RealOrangeOne reviewed (approved):
This looks good to me! Although worth getting a second (third) pair of eyes on it too, just so we're sure!

PeterJCLaw commented:
@sedders123 would you have some time to have a look at this? Would be great to get another set of eyes over it before we merge it.

PeterJCLaw commented:
Ah, looks like this doesn't include the bonus point for leaving the scoring zone.

PeterJCLaw committed:
Move teams data to an instance variable

This will make it easier to modify in tests.

PeterJCLaw committed:
Include the bonus point for leaving the scoring zone

PeterJCLaw commented:
> Ah, looks like this doesn't include the bonus point for leaving the scoring zone.

This is fixed in a5f5320.
